src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-                if(primitive && primitive != MeshPrimitive::Lines) {
-                    Error() << "Trade::ObjImporter::mesh3D(): mixed primitive" << *primitive << "and" << MeshPrimitive::Lines;
-                    return std::nullopt;
-                }
+            int textureId = -1;
+            if(_state->textureIds.find(texture) == _state->textureIds.end()) {
+                /* new texture, create it */
+                int index = _state->textures.size();
+                _state->textures.push_back(texture);
 
-                /* Check vertex count per primitive */
-                if(indexTuples.size() != 2) {
-                    Error() << "Trade::ObjImporter::mesh3D(): wrong index count for line";
-                    return std::nullopt;
-                }
+                _state->textureIds[texture] = index;
+                textureId = index;
+            } else {
+                textureId = _state->textureIds[texture];
+            }
 
-                primitive = MeshPrimitive::Lines;
+            auto suffix = keyword.substr(4, 2);
+            if(suffix == "Kd") {
+                mat->diffuseTexture = textureId;
+            } else if(suffix == "Ka") {
+                mat->ambientTexture = textureId;
+            } else if(suffix == "Ks") {
+                mat->specularTexture = textureId;
+            } else {
+                Warning() << "Trade::ObjImporter::parseMaterialLibrary(): unsupported texture type:" << suffix;
+            }
 
-            /* Faces */
-            } else if(keyword == "f") {
-                /* Check that we don't mix the primitives in one mesh */
-                if(primitive && primitive != MeshPrimitive::Triangles) {
-                    Error() << "Trade::ObjImporter::mesh3D(): mixed primitive" << *primitive << "and" << MeshPrimitive::Triangles;
-                    return std::nullopt;
-                }
+        /* Ignore unsupported keywords, error out on unknown keywords */
+        } else {
+            Warning() << "Trade::ObjImporter::parseMaterialLibrary(): unknown keyword:" << keyword;
+        }
 
-                /* Check vertex count per primitive */
-                if(indexTuples.size() < 3) {
-                    Error() << "Trade::ObjImporter::mesh3D(): wrong index count for triangle";
-                    return std::nullopt;
-                } else if(indexTuples.size() != 3) {
-                    Error() << "Trade::ObjImporter::mesh3D(): polygons are not supported";
-                    return std::nullopt;
-                }
+        /* Ignore the rest of the line */
+        pos = skipWhitespaces(ignoreLine(pos));
+    }
+}
 
-                primitive = MeshPrimitive::Triangles;
+UnsignedInt ObjImporter::doMesh3DCount() const { return _state->meshes.size(); }
 
-            } else CORRADE_ASSERT_UNREACHABLE(); /* LCOV_EXCL_LINE */
+UnsignedInt ObjImporter::doMaterialCount() const { return _state->materials.size(); }
```

Comment:
Just for consistency with other importers I would group all `doMesh*()` functions together, all `doMaterial*()` functions together, etc.